compiler/noirc_evaluator/src/ssa/ir/dfg.rs

```
@@ -194,14 +218,18 @@ impl DataFlowGraph {
         InsertInstructionResult::Results(id, self.instruction_results(id))
     }
 
-    /// Inserts a new instruction at the end of the given block and returns its results
+    /// Simplifies a new instruction and inserts it at the end of the given block and returns its results.
+    /// If the instruction is not handled by the current runtime, `InstructionRemoved` is returned.
     pub(crate) fn insert_instruction_and_results(
         &mut self,
         instruction: Instruction,
         block: BasicBlockId,
         ctrl_typevars: Option<Vec<Type>>,
         call_stack: CallStackId,
     ) -> InsertInstructionResult {
+        if !self.is_handled_by_runtime(&instruction) {
+            return InsertInstructionResult::InstructionRemoved;
+        }
```

This check shouldn't be needed anymore since https://github.com/noir-lang/noir/pull/6894 already prevents inc/dec rc from being inserted into acir functions. They'd only be able to be reintroduced via at inlining optimization pass which inlines brillig code into an acir function, but that shouldn't be possible anyway.

Perhaps we'll have more brillig-only instructions in the future though? (e.g. separating array semantics)?

---

Also (unrelated), with runtime in the dfg now... do we really need both dfg and Function? Function now is just a wrapper around a dfg and an entry block. IIRC we copied this from cranelift originally but I'm unsure of the motivation there. Perhaps Function had more fields.

---

Ah that's true. I didn't notice the circuit diff disappearing from this PR.

---

I saw a runtime check added [here](https://github.com/noir-lang/noir/pull/6894/files#diff-ac8c4aa901333c53b981ac317926a12485d73931bb79c1de9d51a707ba017c9dR472), but AFAICS [constant_folding](https://github.com/noir-lang/noir/blob/ab8807d27b0442bd7e9ff73050f44bacea3add16/compiler/noirc_evaluator/src/ssa/opt/constant_folding.rs#L319) and [loop_invariant](https://github.com/noir-lang/noir/blob/ab8807d27b0442bd7e9ff73050f44bacea3add16/compiler/noirc_evaluator/src/ssa/opt/loop_invariant.rs#L121) can still add them directly. 

FWIW if I return `true` from `is_handled_by_runtime` then we get the following integration test failures because of the panic "Expected all Rc instructions to be removed before acir_gen":
```
    tests::execution_success::test_6::forcebrillig_false_inliner_i64_min_expects
    tests::execution_success::test_array_dynamic_blackbox_input::forcebrillig_false_inliner_i64_min_expects
    tests::execution_success::test_array_to_slice::forcebrillig_false_inliner_i64_min_expects
    tests::execution_success::test_bigint::forcebrillig_false_inliner_i64_min_expects
    tests::execution_success::test_brillig_identity_function::forcebrillig_false_inliner_i64_min_expects
    tests::execution_success::test_conditional_regression_short_circuit::forcebrillig_false_inliner_i64_min_expects
    tests::execution_success::test_debug_logs::forcebrillig_false_inliner_i64_min_expects
    tests::execution_success::test_hashmap::forcebrillig_false_inliner_i64_min_expects
    tests::execution_success::test_hint_black_box::forcebrillig_false_inliner_i64_min_expects
    tests::execution_success::test_merkle_insert::forcebrillig_false_inliner_i64_min_expects
    tests::execution_success::test_nested_array_dynamic::forcebrillig_false_inliner_i64_min_expects
    tests::execution_success::test_pedersen_check::forcebrillig_false_inliner_i64_min_expects
    tests::execution_success::test_pedersen_hash::forcebrillig_false_inliner_i64_min_expects
    tests::execution_success::test_regression_4449::forcebrillig_false_inliner_i64_min_expects
    tests::execution_success::test_sha256::forcebrillig_false_inliner_i64_min_expects
    tests::execution_success::test_sha256_regression::forcebrillig_false_inliner_i64_min_expects
    tests::execution_success::test_sha256_var_padding_regression::forcebrillig_false_inliner_i64_min_expects
    tests::execution_success::test_sha256_var_size_regression::forcebrillig_false_inliner_i64_min_expects
    tests::execution_success::test_sha256_var_witness_const_regression::forcebrillig_false_inliner_i64_min_expects
    tests::execution_success::test_sha2_byte::forcebrillig_false_inliner_i64_min_expects
    tests::execution_success::test_simple_shield::forcebrillig_false_inliner_i64_min_expects
    tests::execution_success::test_slice_dynamic_index::forcebrillig_false_inliner_i64_min_expects
    tests::execution_success::test_slice_regex::forcebrillig_false_inliner_i64_min_expects
    tests::execution_success::test_slices::forcebrillig_false_inliner_i64_min_expects
```

---

> I saw a runtime check added [here](https://github.com/noir-lang/noir/pull/6894/files#diff-ac8c4aa901333c53b981ac317926a12485d73931bb79c1de9d51a707ba017c9dR472), but AFAICS [constant_folding](https://github.com/noir-lang/noir/blob/ab8807d27b0442bd7e9ff73050f44bacea3add16/compiler/noirc_evaluator/src/ssa/opt/constant_folding.rs#L319) and [loop_invariant](https://github.com/noir-lang/noir/blob/ab8807d27b0442bd7e9ff73050f44bacea3add16/compiler/noirc_evaluator/src/ssa/opt/loop_invariant.rs#L121) can still add them directly.

We probably just need to have similar runtime checks in those passes which add the rc instructions directly.

---

Yes, if we add the checks in the passes directly then we can roll back this entire PR. Adding them in the passes might be the only option if the runtime-specific pattern is more complicated than an individual instruction. 

I did it this way because 
1) the ticket suggested there is value in not having instructions which will be just ignored later on and 
2) I thought that the passes that added them could have always been written in a way that they don't add them unless the runtime is meant to process them, so if they weren't, maybe the intention was to keep them agnostic, or at least not more complicated. 

It seemed like silently skipping such instructions was the lightest touch, but I see the value in handling them in the passes as well, saving the extra checks for every instruction that is added.